Add RunSingleTest executable

nilsdeppe opened the pull request:
## Proposed changes

RunSingleTest allows one to compile only one of the many test files and SpECTRE
libraries, and run the tests within the file. This reduces turn around time when
editing low-level headers such as some of the ones found in `Utilities` that are
compiled into the pre-compiled header. The reason is that when the PCH changes
all libraries need to be recompiled, which is very time consuming and not
necessary if only editing something that affects `DataVector`, for example.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add RunSingleTest executable

RunSingleTest allows one to compile only one of the many test files and SpECTRE
libraries, and run the tests within the file. This reduces turn around time when
editing low-level headers such as some of the ones found in `Utilities` that are
compiled into the pre-compiled header. The reason is that when the PCH changes
all libraries need to be recompiled, which is very time consuming and not
necessary if only editing something that affects `DataVector`, for example.

wthrowe commented:
Running `bin/RunSingleTest Unit.DataStructures.DataVector.ref_diff_size` (an assertion test) causes a segfault.  Using RunTests prints the assertion message.

kidder commented:
I believe that is true of RunTest also?

kidder commented:
That may be related to the question of whether or not Parallel::Printf will work or not for #764

nilsdeppe committed:
Use Parallel::printf in ASSERT and ERROR

When not using a Charm++ main function Parallel::abort is broken and results in
a segfault.

nilsdeppe commented:
Okay, I studied this and the segfault is inside `CpdAborting` which is inside `CmiAbort`, which is Charm++ stuff. I messaged them asking for help, but now at least `Parallel::printf` is used to print the ASSERT message so it's clear an assertion failed. I also noted this in the documentation. I think for this testing executable the behavior is fine since the executable will typically be quite small and only used for debugging code.

geoffrey4444 commented on src/ErrorHandling/AbortWithErrorMessage.cpp:
```diff
@@ -22,7 +23,8 @@ void abort_with_error_message(const char* expression, const char* file,
-  Parallel::abort(os.str());
+  Parallel::printf(os.str());
+  Parallel::abort("");
 }
```
Comment that this change enables the error message to print even if the code segfaults, or if I misunderstand, add a comment explaining why this is changed.

geoffrey4444 commented on tests/Unit/RunSingleTest/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+  ${EXECUTABLE}
+  EXCLUDE_FROM_ALL
+  ${EXECUTABLE}.cpp
+  ../DataStructures/Test_DataVector.cpp
```
This example would be clearer to new users if you add a comment above this line, saying that the following lines list the test (or tests, if multiple tests are given) that RunSingleTest can run.

geoffrey4444 commented on docs/DevGuide/WritingTests.md:
```diff
@@ -194,3 +194,20 @@ the `ifdef SPECTRE_DEBUG` block, they can just call have the code that triggers
+changes need to be tested, building `RunTests` becomes extremely time
+consuming. The `RunSingleTest` executable in `tests/Unit/RunSingleTest` allows
+one to compile only a select few of the test source files and only link in the
+necessary libraries. To set which test file and libraries are linked into
```
If it's a select few, and not just one, then shouldn't this be "RunOnlyTheseTests" or something, instead of "RunSingleTest"?

nilsdeppe replied:
The point of the executable is to be able to run a single test without recompiling the entire tree. I'm leaving the name as is.

nilsdeppe commented:
I'm just going to keep using this locally. Feel free to close it.

nilsdeppe commented:
There, take it or leave it.

kidder commented:
@geoffrey4444 @wthrowe please look at

kidder commented:
@wthrowe if you are happy with this, merge it
